File: niffler-e-2-e-tests/src/test/java/guru/qa/niffler/page/MainPage.java
package guru.qa.niffler.page;

import guru.qa.niffler.page.component.Header;
import guru.qa.niffler.page.component.SpendingTable;
import guru.qa.niffler.page.component.StatComponent;
import io.qameta.allure.Step;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import static com.codeborne.selenide.Condition.text;
import static com.codeborne.selenide.Condition.visible;

@ParametersAreNonnullByDefault
public class MainPage extends BasePage<MainPage> {

  public static final String URL = CFG.frontUrl() + "main";

  protected final Header header = new Header();
  protected final SpendingTable spendingTable = new SpendingTable();
  protected final StatComponent statComponent = new StatComponent();

  public Header getHeader() {
    return header;
  }

  @Nonnull
  public StatComponent getStatComponent() {
    return statComponent;
  }

  public SpendingTable getSpendingTable() {
    spendingTable.getSelf().scrollIntoView(true);
    return spendingTable;
  }

  @Step("Check that page is loaded")
  @Override
  public MainPage waitForPageLoaded() {
    header.getSelf().should(visible).shouldHave(text("Niffler"));
    spendingTable.getSelf().should(visible).shouldHave(text("History of Spendings"));
    statComponent.checkThatChartDisplayed();
    return this;
  }
}

File: niffler-e-2-e-tests-python/pages/main_page.py
from selene import browser, be, have
import allure

from pages.base_page import BasePage
from pages.components.header import Header
from pages.components.spending_table import SpendingTable
from utils.config import CFG


class MainPage(BasePage['MainPage']):
    """Main page matching Java MainPage class"""

    URL = f"{CFG.front_url}main"

    def __init__(self):
        super().__init__()
        self._header = Header()
        self._spending_table = SpendingTable()

    @property
    def header(self) -> Header:
        """Get header component"""
        return self._header

    @property
    def spending_table(self) -> SpendingTable:
        """Get spending table component with scroll into view"""
        browser.execute_script("arguments[0].scrollIntoView(true);", self._spending_table.self())
        return self._spending_table

    @allure.step("Check that Main Page is loaded")
    def wait_for_page_loaded(self) -> 'MainPage':
        """Wait for page to be fully loaded"""
        self._header.self.should(be.visible).should(have.text('Niffler'))
        self._spending_table.self.should(be.visible).should(have.text('History of Spendings'))
        self.logger.info("Main Page loaded successfully")
        return self

    @allure.step("Open Main Page")
    def open(self) -> 'MainPage':
        """Open main page"""
        self.logger.info(f"Opening Main Page: {self.URL}")
        browser.open(self.URL)
        return self.wait_for_page_loaded()
